### PR TITLE
Refactor creating PR review comments

### PR DESCRIPTION
**Pull request by 2bndy5**
This abstracts the creation of PR review comments into its own module.

These changes are refactored from initial proposal in #115 but without the new dependency and changes from #116.

### CODE REVIEW

**Comment by 2bndy5** [comment]
Everything seems correct in the [test PR](https://github.com/cpp-linter/test-cpp-linter-action/pull/40).

**Comment by 2bndy5** [comment]
I've got a new idea to merge the suggestions from  both tools... Instead of each tool posting a separate review comment, I might be able to merge the comments from both tools if their line numbers match up. This idea is different from merging the patches from each tool.

**Comment by 2bndy5** [comment]
> if the user only uses clang-format or clang-tidy, they can still work.

Yes

**Comment by 2bndy5** [comment]
Just to illustrate these changes:
## data structures (inheritence)
```mermaid
graph TD
  FileObj -- contains --> TidyAdvice -- inherits from --> PatchMixin
  FileObj -- contains --> FormatAdvice -- inherits from --> PatchMixin
```

## review comment dataflow
### Changes are captured
> [!note]
> The `PatchMixin.patch` member is specific to either `TidyAdvice` or `FormatAdvice` (as described in above inheritance graph).

```mermaid
flowchart TD
  cachePatch["save suggested changes
  to PatchMixin.patched"]
  capTidy["`run_clang_tidy()`"] --> cachePatch
  capFormat["`run_clang_format()`"] --> cachePatch
```

### Review comments are created

Starting from `GithubApiClient.post_feedback()` which will appropriately call `GithubApiClient.post_review()`.

```mermaid
%%{init: {"flowchart": {"htmlLabels": false}} }%%
flowchart TD
  postReviews["post_reviews()"] --> getComments["create_review_comments()"]
  getComments -- if tidy-review is true --> iterFiles
  getComments -- if format-review is true --> iterFiles
  iterFiles["for each file in files"]
  getFmtSuggestion["FormatAdvice
  .get_suggestions_
  from_patch()"]
  getTidySuggestion["TidyAdvice
  .get_suggestions_
  from_patch()"]
  serialize["serialize comments
  into JSON payload"] --> post["post new review"]
  iterFiles --> getFmtSuggestion --> serialize
  iterFiles --> getTidySuggestion --> serialize
```

**Comment by 2bndy5** [comment]
> I've got a new idea to merge the suggestions from both tools...

# Success! :tada:
There's still some edge cases where the suggestion comments cannot be merged, but I'm happier with the new results (compared to old/stable results).

There's also a couple lines that get missed in coverage. This is because the test conditions **do not include** a scenario where a clang-tidy diagnostic comment (with no suggested fix -- see below) **cannot** be merged into existing review comments.
image
I'm ok with the loss of coverage in this situation (for now).

**Comment by 2bndy5** [comment]
I realized I keep using dense/complex sentences (with double negatives), and the code is not super simple. That's why I created the graphs and picture.

File: pyproject.toml [hunk 1]
```diff
@@ -31,7 +26,7 @@ classifiers = [
     "Programming Language :: Python :: 3",
     "Topic :: Software Development :: Build Tools",
 ]
-dynamic = ["version"]
+dynamic = ["version", "dependencies"]
```
[the remark this answers is not in the document]
Comment (2bndy5) [reply pyproject.toml:29]: Good catch! I'll update it.

File: cspell.config.yml
[the remark this answers is not in the document]
Comment (2bndy5) [reply cspell.config.yml:1]: The [codespell tool](https://github.com/codespell-project/codespell) is different from [cspell tool](https://github.com/streetsidesoftware/cspell). codespell ~is the only tool I could find that~ had a pre-commit hook. Locally, I use the [cspell VSCode extension](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) for real-time corrections (includes checking git commit messages as you type them).

The fact that codespell doesn't pick up on all the exceptions in this cspell config means **cspell is more efficient than codespell**. It might be worth switching to [cspell-cli as the pre-commit hook](https://github.com/streetsidesoftware/cspell-cli#setup-pre-commit-hook).

[the remark this answers is not in the document]
Comment (2bndy5) [reply cspell.config.yml:1]: changed pre-commit hook to cspell-cli
